paddle/fluid/distributed/auto_parallel/spmd_rules/layer_norm_spmd_rule.cc

```diff
@@ -0,0 +1,163 @@
+
+namespace paddle {
+namespace distributed {
+namespace auto_parallel {
+using phi::distributed::auto_parallel::str_join;
+std::pair<std::vector<TensorDistAttr>, std::vector<TensorDistAttr>>
+LayerNormSPMDRule::InferForward(const std::vector<DistTensorSpec>& input_specs,
+                                const paddle::framework::AttributeMap& attrs) {
+  // step0: verify input args based on matmul logic
+  auto input_specs_size = input_specs.size();
+  PADDLE_ENFORCE_EQ(
+      input_specs_size,
+      3,
+      phi::errors::InvalidArgument(
+          "The size of InputSpec of layer_norm should be 3, but got [%d].",
+          input_specs_size));
+  auto x_shape = input_specs[0].shape();
+  auto scale_shape = input_specs[1].shape();
+  auto bias_shape = input_specs[2].shape();
+  int x_ndim = x_shape.size();
+  int scale_ndim = scale_shape.size();
+  int bias_ndim = bias_shape.size();
+
+  PADDLE_ENFORCE_EQ(
+      scale_ndim,
+      1,
+      phi::errors::InvalidArgument(
+          "The ndim of scale in layer_norm should be 1, but got [%d].",
+          scale_ndim));
+
+  PADDLE_ENFORCE_EQ(
+      bias_ndim,
+      1,
+      phi::errors::InvalidArgument(
+          "The ndim of bias in layer_norm should be 1, but got [%d].",
+          bias_ndim));
+
+  auto x_dims_mapping = input_specs[0].dist_attr().dims_mapping();
+  auto scale_dims_mapping = input_specs[1].dist_attr().dims_mapping();
+  auto bias_dims_mapping = input_specs[2].dist_attr().dims_mapping();
+
+  auto x_dist_attr_src = input_specs[0].dist_attr();
+
+  std::vector<TensorDistAttr> input_dist_attrs;
+  input_dist_attrs.reserve(input_specs.size());
+
+  int begin_norm_axis = ExtractAttr<int>("begin_norm_axis", attrs);
+
+  // Step2.3.2  handle input tensor partial (TODO)
+  VLOG(4) << "LayerNormSPMDRule InferForward Inputs: "
+          << "x shape: [" << str_join(x_shape) << "], x_dims_mapping: ["
+          << str_join(x_dims_mapping) << "]; scale shape: ["
+          << str_join(scale_shape) << "], scale_dims_mapping: ["
+          << str_join(scale_dims_mapping) << "]; bias shape: ["
+          << str_join(bias_shape) << "], bias_dims_mapping: ["
+          << str_join(bias_dims_mapping) << "]; begin_norm_axis: ["
+          << begin_norm_axis << "]; ";
+
+  // step1: build Einsum Notation
+  // ijk,k,k->ijk (x,scale,bias->out, begin_norm_axis=2)
+  // ijkl,z(kl),z(kl)->ijkl (x,scale,bias->out, begin_norm_axis=2, z=kl)
+  std::string x_axes = "";
+  for (auto i = 0; i < x_ndim; ++i) {
+    x_axes += static_cast<char>(static_cast<int>('k') - begin_norm_axis + i);
+  }
+
+  std::string scale_axes;
+  std::string bias_axes;
+  if (x_ndim - begin_norm_axis == 1) {
+    scale_axes = "k";
+    bias_axes = "k";
+  } else {
+    // z = x_axes.substr(begin_norm_axis, x_ndim - begin_norm_axis)
+    scale_axes = "z";
+    bias_axes = "z";
+  }
+  std::string out_axes = x_axes;
+
+  VLOG(4) << "LayerNormSPMDRule build Einsum notation (x,scale,bias->out): ["
```

Comment:
should include all layer_norm outputs (x,scale,bias->out, mean, var)

Reply:
done, thx